src/ITensorNetworks/itensor_network.jl

```diff
@@ -0,0 +1,201 @@
+the local state `state`.
+"""
+function boundary_projectors(tn::Matrix{ITensor}, state=1)
+  top_row = tn[1, :]
+  bottom_row = tn[end, :]
+  left_column = tn[:, 1]
+  right_column = tn[:, end]
+  bottom_boundary_inds = commonind.(bottom_row, top_row)
+  right_boundary_inds = commonind.(right_column, left_column)
+  ψr = ITensors.state.(right_boundary_inds, state)
+  ψb = ITensors.state.(bottom_boundary_inds, state)
+  return ψr, ψb
+end
+
+"""
+    project_boundary(tn::Matrix{ITensor}, state=1)
+
+Project the boundary of a periodic 2D tensor network onto 
+the specified state.
+"""
+function project_boundary(tn::Matrix{ITensor}, state=1)
+  Nx, Ny = size(tn)
+  ψr, ψb = boundary_projectors(tn, state)
+  for n in 1:Nx
+    tn[n, 1] = tn[n, 1] * ψr[n]
+    tn[n, end] = tn[n, end] * dag(ψr[n])
+  end
+  for n in 1:Ny
+    tn[1, n] = tn[1, n] * ψb[n]
+    tn[end, n] = tn[end, n] * dag(ψb[n])
+  end
+  return tn
+end
+
+function filter_alllinkinds(f, tn)
+  linkinds = Dict{Tuple{keytype(tn),keytype(tn)},Vector{indtype(tn)}}()
+  for n in keys(tn), m in keys(tn)
+    if f(n, m)
+      is = commoninds(tn[n], tn[m])
+      if !isempty(is)
+        linkinds[(n, m)] = is
+      end
+    end
+  end
+  return linkinds
+end
+
+"""
+    alllinkinds(tn)
+
+Return a dictionary of all of the link indices of the network.
+The link indices are determined by searching through the network
+for tensors with indices in common with other tensors, and
+the keys of the dictionary store a tuple of the sites with
+the common indices.
+
+Notice that this version will return a dictionary containing
+repeated link indices, since the 
+For example:
+```julia
+i, j, k, l = Index.((2, 2, 2, 2))
+inds_network = [(i, dag(j)), (j, dag(k)), (k, dag(l)), (l, dag(i))]
+tn_network = randomITensor.(inds_network)
+links = allinkinds(tn_network)
+links[(1, 2)] == (dag(j),)
+links[(2, 1)] == (j,)
+links[(2, 3)] == (dag(l),)
+links[(1, 3)] # Error! In the future this may return an empty Tuple
+```
+
+Use `inlinkinds` and `outlinkinds` to return dictionaries without
+repeats (such as only the link `(2, 1)` and not `(1, 2)` or vice versa).
+"""
+alllinkinds(tn) = filter_alllinkinds(≠, tn)
+inlinkinds(tn) = filter_alllinkinds(>, tn)
+outlinkinds(tn) = filter_alllinkinds(<, tn)
+
+function filterneighbors(f, tn, n)
+  neighbors_tn = keytype(tn)[]
+  tnₙ = tn[n]
+  for m in keys(tn)
+    if f(n, m) && hascommoninds(tnₙ, tn[m])
+      push!(neighbors_tn, m)
+    end
+  end
+  return neighbors_tn
+end
+
+"""
+    neighbors(tn, n)
+
+From a tensor network `tn` and a site/node `n`, determine the neighbors
+of the specified tensor `tn[n]` by searching for which other
+tensors in the network have indices in common with `tn[n]`.
+
+Use `inneighbors` and `outneighbors` for directed versions.
+"""
+neighbors(tn, n) = filterneighbors(≠, tn, n)
+inneighbors(tn, n) = filterneighbors(>, tn, n)
+outneighbors(tn, n) = filterneighbors(<, tn, n)
+
+function mapinds(f, ::typeof(linkinds), tn)
```

Comment:
I'm curious why do we need a ::typeof(linkinds) input here, this is not used in the function?

Reply:
It is to distinguish between three different sets of functionality:

1. `mapinds(f, tn)` (for example `prime(tn)`) to prime all indices.
2. `mapinds(f, ::typeof(linkinds), tn)` (for example `prime(linkinds, tn)`) to prime/tag all internal indices of the network.
3. `mapinds(f, ::typeof(siteinds), tn)` (for example `prime(siteinds, tn)`) to prime/tag all external indices of the network.

So it is used for purpose of dispatch, and not necessarily needed in the function itself. In ITensors.jl, I started out by calling this `mapinds(f, tn)`, `mapinds_linkinds(f, tn)`, and `mapinds_siteinds(f, tn)` but I found that the function names were long and cumbersome, and for the sake of code organization it was easier to pass information about which subset of indices were being modified as an argument.

The interpretation I have for it is that the argument signifies a certain subset of the indices that are being modified, so `prime(linkinds, tn)` primes all indices that result from the call `linkinds(tn)` (it could be implemented in this way, but isn't right now).